### DISCUSSION TITLE
Render different components based on screen size in Next.js

### DISCUSSION THREAD

**Discussion by theairbend3r** [vercel/next.js#14810] [Help]
I'm trying to render render a component differently based on if the screen width < 768p. 

If it is less than 768p, then load the hamburger menu. If not, load the entire menu.

Here's the code I use. The issue is, when the page first loads, I cannot observe the changes. But when I make the browser screen small and then increase it to it's original size; the effects take place. 

I think it's because React is being rendered server-side. But still don't understand why it works by making the screen smaller and then bigger again. 

```
import "twin.macro"
import { Global, css } from "@emotion/core"

import { useState, useEffect } from "react"

const Navbar = () => {
  const useWindowDimensions = () => {
    const hasWindow = typeof window !== "undefined"

    function getWindowDimensions() {
      const width = hasWindow ? window.innerWidth : null
      const height = hasWindow ? window.innerHeight : null
      return {
        width,
        height,
      }
    }

    const [windowDimensions, setWindowDimensions] = useState(
      getWindowDimensions()
    )

    useEffect(() => {
      if (hasWindow) {
        function handleResize() {
          setWindowDimensions(getWindowDimensions())
        }

        window.addEventListener("resize", handleResize)
        return () => window.removeEventListener("resize", handleResize)
      }
    }, [hasWindow])

    return windowDimensions
  }

  const { height, width } = useWindowDimensions()
  const breakpoint = 768

  return (
    <div>
      {width <= breakpoint ? (
        <div>
          <HamburgerMenu />
        </div>
      ) : (
        <div>
           <FullMenu />
        </div>
  )
}

export default Navbar

```

Is there a way I can force the Next.js app to apply the effects on the first render?

**Comment by elchaparrito1** [comment]
I'm in the exact same situation, and would love to see a response posted here. I need one of my Next.js components to know the innerWidth of the window object on the initial render.

**Comment by jrgarciadev** [comment]
Hi, you can use a media query hook

```
import { useState, useCallback, useEffect } from 'react';

const useMediaQuery = (width) => {
  const [targetReached, setTargetReached] = useState(false);

  const updateTarget = useCallback((e) => {
    if (e.matches) {
      setTargetReached(true);
    } else {
      setTargetReached(false);
    }
  }, []);

  useEffect(() => {
    const media = window.matchMedia(`(max-width: ${width}px)`);
    media.addListener(updateTarget);

    // Check on mount (callback is not called until a change occurs)
    if (media.matches) {
      setTargetReached(true);
    }

    return () => media.removeListener(updateTarget);
  }, []);

  return targetReached;
};


const Navbar = () => {
   const isBreakpoint = useMediaQuery(768)
   return (
    <div>
      { isBreakpoint ? (
        <div>
          <HamburgerMenu />
        </div>
      ) : (
        <div>
           <FullMenu />
        </div>
   )
)}

export default Navbar;
```

**Reply by srithan** [reply]
Hey @jrgarciadev. Thank you for this. This works right good.
Do you have any updates to this code?

**Reply by jajibhee** [reply]
```
useEffect(() => {
    const media = window.matchMedia(`(max-width: ${width}px)`)
    media.addEventListener('change', e => updateTarget(e))

    // Check on mount (callback is not called until a change occurs)
    if (media.matches) {
      setTargetReached(true)
    }

    return () => media.removeEventListener('change', e => updateTarget(e))
  }, [])
```

Incase it throws the addListener or removeListener deprecated error.

**Reply by allan2** [reply]
Shouldn't the dependencies `updateTarget` and `width` be included in the dependency array?

From the `react-hooks/exhaustive-deps` lint rule

**Reply by buqento** [reply]
nice share. thanks dev

bvqento richard
tantekos.com

**Reply by brianlfarmerllc** [reply]
@jrgarciadev Thanks! This worked great for me in my Next.js project in the two browsers I've tested so far, Chrome and Firefox.

**Reply by edwinvelez** [reply]
Hey @jrgarciadev thanks for the code. It works as is expected. Great job!

**Reply by builtbynil** [reply]
thats awesome.
but i can not understand how this code works on the server!
could you pls explain me what is happening?

here is my code: 
import { useCallback, useEffect, useState } from 'react';
import { TdisplaySizes } from 'models';

const useMediaQuery = minWidth => {
    const [targetReached, setTargetReached] = useState(false);
    const updateTarget = useCallback(e => setTargetReached(!!e.matches), []);

    useEffect(() => {
        const handler = e => updateTarget(e);
        const media = window.matchMedia(`(min-width: ${minWidth}px)`);
        media.addEventListener('change', handler);
        // Check on mount (callback is not called until a change occurs)
        if (media.matches) setTargetReached(true);
        return () => media.removeEventListener('change', handler);
    }, [])

    return targetReached;
};

const useResponsive = () => {
    const displaySize: TdisplaySizes = {
        xs: useMediaQuery('0'),
        sm: useMediaQuery('640'),
        md: useMediaQuery('768'),
        lg: useMediaQuery('1024'),
        xl: useMediaQuery('1280'),
        twoXl: useMediaQuery('1536'),
    }
    return displaySize;
}

export default useResponsive;

**Reply by TakshakRamteke** [reply]
It worked,
Thankyou So much :purple_heart: 

> Hi, you can use a media query hook
> 
> ```
> import { useState, useCallback, useEffect } from 'react';
> 
> const useMediaQuery = (width) => {
>   const [targetReached, setTargetReached] = useState(false);
> 
>   const updateTarget = useCallback((e) => {
>     if (e.matches) {
>       setTargetReached(true);
>     } else {
>       setTargetReached(false);
>     }
>   }, []);
> 
>   useEffect(() => {
>     const media = window.matchMedia(`(max-width: ${width}px)`);
>     media.addListener(updateTarget);
> 
>     // Check on mount (callback is not called until a change occurs)
>     if (media.matches) {
>       setTargetReached(true);
>     }
> 
>     return () => media.removeListener(updateTarget);
>   }, []);
> 
>   return targetReached;
> };
> 
> 
> const Navbar = () => {
>    const isBreakpoint = useMediaQuery(768)
>    return (
>     <div>
>       { isBreakpoint ? (
>         <div>
>           <HamburgerMenu />
>         </div>
>       ) : (
>         <div>
>            <FullMenu />
>         </div>
>    )
> )}
> 
> export default Navbar;
> ```

**Reply by Negar-Arianmehr** [reply]
**Thank you**, it is working.

Just for somebody who wants to use it now, I got the warning for removeListener and addListener

> MediaQueryList.addListener()
Deprecated: This feature is no longer recommended. Though some browsers might still support it, it may have already been removed from the relevant web standards, maybe in the process of being dropped, or may only be kept for compatibility purposes. Avoid using it, and update existing code if possible; see the compatibility table at the bottom of this page to guide your decision. Be aware that this feature may cease to work at any time.

see: 
1. [https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList/addListener](url)
2. [https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList/removeListener](url)

I replaced  
`media.addListener(updateTarget);  ` 
with
`       media.addEventListener("change", updateTarget)
`
and `return () => media.removeListener(updateTarget);`
with
` return () => media.removeEventListener("change", updateTarget);`

**Reply by zhiln** [reply]
Works perfectly smooth indeed!

Was anybody able to test this implementation with Jest and React Testing Library? Mocking the matchMedia isn't a problem, but making the execution during the test to reach `updateTarget` function is what I'm struggling with a whole day. Any ideas? 🙏

**Reply by sbgprojects** [reply]
Great, Thanks a lot. I've been looking for the solution for a long time.

**Reply by eltel** [reply]
> { isBreakpoint ? (
>         <div>
>           <HamburgerMenu />
>         </div>
>       ) : (
>         <div>
>            <FullMenu />
>         </div>
>    )
> )}

This (with the updates from @Negar-Arianmehr)  has helped me in oh so many ways, many thanks guys!!

**Reply by n8slo88** [reply]
Awesome. Yes. I couldn't compile as addEventListener was looking for 2 arguments.  I had to add "change".  thank you.

**Reply by dlsaldanas** [reply]
just a little update
```
const useMediaQuery = (width) => {
  const [targetReached, setTargetReached] = useState(false);

  const updateTarget = useCallback((e) => {
    if (e.matches) {
      setTargetReached(true);
    } else {
      setTargetReached(false);
    }
  }, []);

  useEffect(() => {
    const media = window.matchMedia(`(max-width: ${width}px)`);
    if (media.addEventListener) {
      media.addEventListener("change", updateTarget);
    } else {
      // compatibility for browser that dont have addEventListener
      media.addListener(updateTarget);
    }
    // Check on mount (callback is not called until a change occurs)
    if (media.matches) {
      setTargetReached(true);
    }
    if (media.removeEventListener) {
      return () => media.removeEventListener('change', updateTarget);
    } else {
      // compatibility for browser that dont have removeEventListener
      return () => media.removeListener(updateTarget);
    }
  }, []);

  return targetReached;
};
```

**Reply by devDanielCespedes** [reply]
I love you

**Reply by rishabh1S** [reply]
Here is an update for useMediaQuery for the latest versions.


```
const useMediaQuery = (width: number) => {
  const [targetReached, setTargetReached] = useState(false);

  const updateTarget = useCallback((e: MediaQueryListEvent) => {
    if (e.matches) {
      setTargetReached(true);
    } else {
      setTargetReached(false);
    }
  }, []);

  useEffect(() => {
    const media = window.matchMedia(`(max-width: ${width}px)`);
    media.addEventListener("change", updateTarget);

    if (media.matches) {
      setTargetReached(true);
    }

    return () => media.removeEventListener("change", updateTarget);
  }, []);

  return targetReached;
};
```

**Reply by ygorbunkov** [reply]
Regardless of the media query string passed to `window.matchMedia().matches` it cannot be evaluated to `true` on the server (as `window` is `undefined`). 

This will result in default layout (let it be mobile) rendered server side and then (once React kicks in) reflow on the client (which may happen to be desktop). 

Under condition of slow internet connection it may appear to user as content jitter similar to [FOUC](https://en.wikipedia.org/wiki/Flash_of_unstyled_content) causing [layout shift](https://web.dev/articles/cls) and subsequently hitting web app performance and overall UX badly.

One possible solution of that problem could be based on [client hint header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-CH-UA-Mobile) which is [not yet](https://caniuse.com/?search=Sec-CH-UA-Mobile) part of the standard by now.

**Comment by jonrrivera** [comment]
amazing will try this out. thanks for the help

- Jon Rivera
jonrrivera.com

**Reply by ayobamiseun** [reply]
do we need to "use client"

**Reply by estate-market-admin** [reply]
Using Dynamic import should do the trick 



`'use client'
 
import { useState } from 'react'
import dynamic from 'next/dynamic'
 
// Client Components:
const ComponentA = dynamic(() => import('../components/A'))
const ComponentB = dynamic(() => import('../components/B'))
 
export default function ClientComponentExample() {
  const [isMobile, setIsMobile] = useState(false)
 
 // manage which viewport somewhere here
 
  return (
    <div>
      {/* Only load either component A or B depending on viewport  */}
    {  isMobile ? <ComponentA />  : <ComponentB />} }
 
  )
}
`

**Comment by bennik88** [comment]
So does this code runs serverside and therefore preserve static or serverside rendered pages? I thought using useEffect would mean you opt out of serverside rendering?

**Reply by mattapperson** [reply]
This means that the SSR will always return false from this hook

**Reply by ygorbunkov** [reply]
You seem to be confusing RSC with SSR. Client components (making use of media queries to offer responsive UI) may still be rendered on the server while resulting HTML will correspond to `window.matchMedia().matches` evaluated to `false`  and then reflow (if necessary) will happen on the client. You're right though, in order for `window.matchMedia()` to ever result into something else than reference error, it has to be client component.

**Comment by bdombro** [comment]
The proposed solution is very flawed since it causes mismatches between server-render and client-render. I'd love to see an official solution for this.

**Reply by Sebastp** [reply]
+1

**Comment by BhavyaCodes** [comment]
I think the only way to do this would be to get this data from nextjs app context, more specifically the `User-Agent` request header to somewhat detect the device type, I don't know about app router, Im only talking in context of getServerSideProps or getInitialProps

**Comment by MrProgrammer** [comment]
it could be OK on server component for SSR

import { headers } from 'next/headers'

export const getDeviceType = () => {
  const headersList = headers()
  const userAgent = headersList.get('user-agent')

  return userAgent!.match(
    /Android|BlackBerry|iPhone|iPad|iPod|Opera Mini|IEMobile|WPDesktop/i
  )
    ? 'mobile'
    : 'desktop'
}

**Reply by mdjfs** [reply]
you're a genius. Thank you !

**Reply by ygorbunkov** [reply]
`User-Agent` header is not quite reliable (as it can be spoofed by certain browsers). Furthermore, this approach doesn't scale too well as you won't be able to make a difference between tablet and mobile which you might require.

**Comment by nil1511** [comment]
Use [Sec-CH-UA-Mobile client hint](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-CH-UA-Mobile) for server-side rendering for mobile and desktop, and fallback to user-agent for Safari. There is also a `Viewport-Width` client hint which gives width on the server but that is deprecated

**Comment by NakZMichael** [comment]
Can we cache rendered result separately for desktop and mobile if we use `user-agent`?

**Reply by MrProgrammer** [reply]
we use user-agent for not render mobile component on desktop size and vice versa
it's not ok we render mobile component on dektop size!

**Comment by Athar-Naveed** [comment]


**Reply by sFritsch09** [reply]
I do the same like @Athar-Naveed but I use screen.width instead:
```javascript
...
setWidth(screen.width);
...
```

**Reply by ygorbunkov** [reply]
The approach based on `window.matchMedia()` seems to be much cleaner as it doesn't require any handcrafting to keep track of the screen width on resize. 

And just like media queries, `window.innerWidth` will result into certain default layout rendered on the server (where no `innerWidth` exists) and then (once React kicks in on the client) trigger reflow, which will hit badly UX in general and [CLS](https://web.dev/articles/cls), in particular.

Furthermore, leaning onto `addEventListener()` would require debounce not to overwhelm the client by triggering state updates per every pixel of width change while stretching the window.

**Comment by impsdc** [comment]
Hi guys, I'm trying to achieve this in Next 14 app router in middleware.ts. I use [userAgent](https://nextjs.org/docs/app/api-reference/functions/userAgent) to retrieve the device information but it always undefined, Is this normal ?
When I retrieve "user-agent" from the headers, I get a string but I can't find a way to determine the client window.width using it.
``` 
import { headers } from "next/headers";
import { type NextRequest, userAgent } from "next/server";

export function middleware(request: NextRequest) {
	const { device } = userAgent(request);  // undefined
	const headersList = headers();
	const usrAgent = headersList.get("user-agent"); // Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36
}
```

Any help ?

**Reply by ygorbunkov** [reply]
There's no way to find out the exact screen width using `User-Agent` header. Moreover, this particular header may be spoofed (for anonymity purpose) by certain browsers.

**Reply by ibanlopez** [reply]
> There's no way to find out the exact screen width using `User-Agent` header. Moreover, this particular header may be spoofed (for anonymity purpose) by certain browsers.

https://developer.chrome.com/docs/privacy-security/user-agent-client-hints

**Reply by ygorbunkov** [reply]
It still has relatively low browser support, though.
https://caniuse.com/?search=client-hints

**Reply by ibanlopez** [reply]
I've playing around with react use() hook.
You can render server side a default assumption, pass down the promise to the client component, the client component resolves the promise when knows the screen size, then gracefully applies and adjusts the render to the client "real" screen size.

Demo: [https://github.com/ibanlopez/v0-dynamic-rendering-issue](https://github.com/ibanlopez/v0-dynamic-rendering-issue)

**Comment by ArshlanSiddique1** [comment]
It's work for me on next.js 14.

------------------- Custom Hook for Getting MediaQuery , Device Type or Width for Next.js 14. ------------------------------
import { useState, useEffect } from 'react';

const useMediaQuery = (query) => {
    const [matches, setMatches] = useState(false);

    useEffect(() => {
        const media = window.matchMedia(query);
        if (media.matches) {
            setMatches(true);
        }

        const listener = (event) => setMatches(event.matches);
        media.addEventListener('change', listener);

        return () => media.removeEventListener('change', listener);
    }, [query]);

    return matches;
};

export default useMediaQuery;    

---------------------------- Use cases on component ---------------------------
import useMediaQuery from './useMediaQuery'; 

const isMobile = useMediaQuery('(max-width: 768px)');
const isTablet = useMediaQuery('(min-width: 768px) and (max-width: 992px)');
// ... other media queries

return (
    <div>
        {isMobile && <p>This content is displayed on mobile devices.</p>}
    </div>
);

**Comment by ibanlopez** [comment]
This is what I use:

I define a custom hook: `use-media-query.ts` in a `lib` folder:

**src/lib/hooks/use-media-query.ts**:

```
'use client'

import { useState, useEffect } from 'react'

export default function useMediaQuery(query: string): boolean {

    const [matches, setMatches] = useState(false)

    useEffect(() => {
        const media = window.matchMedia(query)

        if (media.matches !== matches) {
            setMatches(media.matches)
        }
        const listener = () => {
            setMatches(media.matches)
        }
        media.addEventListener('change', listener)
        return () => media.removeEventListener('change', listener)
    }, [matches, query])

    return matches
}
```

Then I create a client component to deal with the browser, but most importantly, to be able to render RSC components:

**src/components/composables/media-rendering.tsx**:

```
'use client'

import useMediaQuery from '@/lib/hooks/use-media-query'

const MediaRendering = ({
    minWidth, maxWidth, children
}: {
    minWidth: string | null,
    maxWidth: string | null,
    children: React.ReactNode
}) => {

    const minDeviceSize = useMediaQuery(`(min-width: ${minWidth}px)`)
    const maxDeviceSize = useMediaQuery(`(max-width: ${maxWidth}px)`)

    return (
        <>{(maxDeviceSize || minDeviceSize) && children}</>
    )
}

export default MediaRendering
```

So, let's suppose I have a `navbar.tsx` that I need to render differently for mobile and desktop:

```
import MediaRendering from '@/components/composables/media-rendering'
import NavMenu from './nav-menu'
import Logo from './logo'
import Hamburger from './hamburger-button'

export default function Navbar() {
    const menuLinks = // Do the fetch for links here.
    return (
        <nav className="absolute top-0 z-50 left-0 w-full">
            <div className="relative bg-transparent w-full p-4 flex flex-row flex-wrap items-center md:py-4 md:px-10 lg:flex-nowrap lg:p-4 xl:max-w-screen-xl xl:mx-auto z-30 xlnp:px-0">
                <Logo />
                <MediaRendering minWidth={null} maxWidth="1023">
                    <Hamburger>
                        <NavMenu menuLinks={menuLinks} />
                    </Hamburger>
                </MediaRendering>
                <MediaRendering minWidth="1024" maxWidth={null}>
                    <NavMenu menuLinks={menuLinks} />
                </MediaRendering>
            </div>
        </nav>
    )
}
```

Just FYI: this would be the **NavMenu.tsx**:

```
import Link from 'next/link'

export default function NavMenu({ menuLinks }: { menuLinks: {
    id: number,
    link: string,
    text: string,
    title: string
}[] }) {

    const renderedLinks = menuLinks.map(link => {
        return (
            <li key={link.id} className="flex-initial w-full text-white border-b border-b-white last:border-b-0 hover:bg-yellow-600 lg:text-yellow-400 lg:hover:bg-transparent lg:w-auto lg:border-b-0">
                <Link
                    href={link.link}
                    title={link.title}
                    className="px-4 py-2 block w-full font-semibold antialiased"
                >
                    {link.text}
                </Link>
            </li>
        )
    })

    return (
        <ul className="w-full flex flex-row flex-wrap rounded-md bg-yellow-500 lg:grow lg:flex-nowrap lg:bg-transparent lg:justify-end xl:w-[600px]">
            {renderedLinks}
        </ul>
    )
}
```

**Reply by Mainak908** [reply]
Thanks bro.It worked for me.thnx a lot.I was exactly looking for this solution

**Reply by jsefiani** [reply]
Hey @ibanlopez , what do you mean by:

> but most importantly, to be able to render RSC components

Is it because you can pass a server component to `MediaRendering` as children?

Why don't you use something like react-responsive?

**Reply by jsefiani** [reply]
Using this approach won't prevent the layout from flickering 😓

**Reply by luiscaceresd** [reply]
thanks! crack

**Comment by ibanlopez** [comment]
> react-responsive



> Hey @ibanlopez , what do you mean by:
> 
> > but most importantly, to be able to render RSC components
> 
> Is it because you can pass a server component to `MediaRendering` as children?
> 
> Why don't you use something like react-responsive?

Sorry for the delay in my answer. I usually don't need so much stuff in order to achieve this. Performance for me is key.

And yes, I pass a React server Component as the children of the `MediaRendering`.

**Comment by ibanlopez** [comment]
> Using this approach won't prevent the layout from flickering 😓

If by "flickering" you mean that you experience shift arounds that affect CLS, that's out of the scope of the way you use to render something different from another thing.

Either way you use it's going to cause "flickering" if the `height` of those components are different. You fix this with CSS, skeletons, or even there's no possible "flickering" if you load it once, which is the purpose of this method. But let's say you are in a tablet device, the orientation would cause a re-render: in this case you have to control the `heigh`.

**Comment by ibanlopez** [comment]
For the record: 

You can go just with one `minWidth` or `maxWidth` prop, there's no need to be so verbose. So you either render or not based in one prop only, the best it suits your goal over the design.

In this particular case I had to use this `MediaRendering` twice in the component because there's a difference in what I want to render inside them, but that does not happen so often.

Normally you render something or not at all based on the device width, so you only need a prop.

**Comment by Jordaneisenburger** [comment]
Just chiming in here would also love to see a real Next.js way of doing this. Using matchMedia isn't a good solution as it causes CLS/FOUC if the elements are below the fold it's fine but otherwise it's bad.

**Comment by mjbalcueva** [comment]
[Here's](https://gist.github.com/mjbalcueva/a875538b9dac92a3e660e031652bb3dc) one that persist its state using cookies to avoid the small flickering when reloading.

**Reply by ibanlopez** [reply]
Bad idea. In terms of performance that's terrible, wrapping your rootLayout with [dynamic functions](https://nextjs.org/docs/app/building-your-application/rendering/server-components#dynamic-functions) like `headers`, `cookies` o `searchParams` is a no no go for me.

**Reply by mjbalcueva** [reply]
Thanks for spotting that!

**Reply by ibanlopez** [reply]
Try to SSG or ISR and see if the cache keeps the pages. Run a build and then start. After that see the response headers on the network tab of the browser. 🚀

**Reply by ibanlopez** [reply]
Demo: [https://github.com/ibanlopez/v0-dynamic-rendering-issue](https://github.com/ibanlopez/v0-dynamic-rendering-issue).

Just playing with React use() hook.

**Comment by avonow** [comment]
I've solved my "version" of this issue purely in CSS. I always render both the mobile and the desktop version of the component, but I apply a default `display: hidden` to the elements with an `@media` query. 

This means I avoid the need for a screen size hook entirely and I also don't get the loading "stretch". This also allows me to serve my toolbar as an RSC, since it doesn't need any client hooks.

**Reply by weizhe412** [reply]
yep it's a way, but the the component itself is still in the dom tree

**Reply by ygorbunkov** [reply]
You're supposed to pay twice for such decision in terms of performance:
- first, when you need to load (run) extra JavaScript from your bundle (for both desktop and mobile)
- second, when browser needs to deal with the bloated DOM (with part of it hidden)

**Reply by ElectricCodeGuy** [reply]
"bloated DOM" 
If you do it right using display: 'none' a small amount of extra javascript in the bundle is always preferred over using a hook causing layout shift. Especially for users who have a bad network connection this hooks can really be a dealbreaker for some users. A user is not gonna notice a small amount of extra HTML in the browser he can't even see.  A user is gonna notice a layout shift on his mobile phone. 

Using CSS to do CSS stuff is the way to go.
